engine/runner/src/main/java/org/enso/runner/Main.java

```diff
@@ -1459,6 +1466,80 @@ final CommandLine preprocessArguments(String... args) {
     }
   }
 
+  /**
+   * This method has to be called as early as possible. It attempts to find the project root
+   * directory of the given file, and if the project root is found, it uses native code to change
+   * the working directory to the project root. In order for the JVM's {@code java.io} to reflect
+   * the working directory change, this methods must be called before any class from {@code java.io}
+   * is accessed.
+   *
+   * <p>Note that invoking native code is the only reliable way to change the working directory in
+   * the current process.
+   *
+   * <p>For detailed explanation see this <a
+   * href="https://github.com/enso-org/enso/pull/12618#issuecomment-2778451448">GH comment</a>.
+   *
+   * @param fileToRun the file to run, value of the {@code --run} option.
+   */
+  private void maybeChangeWorkingDirToProjectRoot(String fileToRun) {
+    assert fileToRun != null;
+    if (!ImageInfo.inImageRuntimeCode()) {
+      logger.warn("Changing working directory only supported in native image");
+      return;
+    }
+    var projectRoot = findProjectRoot(fileToRun);
+    var nativeApi = WorkingDirectories.getCurrent();
+    if (projectRoot != null) {
+      var parentDir = parentFile(projectRoot);
+      assert parentDir != null;
+      var curDir = nativeApi.currentWorkingDir();
+      if (!parentDir.equals(curDir)) {
+        var dirChanged = nativeApi.changeWorkingDir(parentDir);
+        if (!dirChanged) {
+          logger.error("Cannot change working directory to {}", parentDir);
+        }
+      }
+    }
+  }
+
+  /**
+   * Attempts to find project root directory. Does not use anything from {@code java.io} on purpose.
+   *
+   * @return null if project root was not found, a canonical path otherwise.
+   */
+  private static String findProjectRoot(String path) {
+    var nativeApi = WorkingDirectories.getCurrent();
+    // Cannot use File.separator char here
+    var separatorChar =
```

Comment:
Unused variable.

Reply:
Removed in 06258375af8e3538cf151b116351b2af0872448e